### ISSUE TITLE
Create Content Block paragraph for Mini Landing Pages

### ISSUE THREAD

**Issue by sarinapadilla** [NCIOCPL/cgov-digital-platform#4359]
## As a content editor, I want to be able to add a content block to my mini landing page, so that I can feature custom content.

* NCIDS Streamlined WYSIWYG: headings (`h1`-`h6`), inline styles (bold, italics, strikethrough, superscript, subscript), remove format, ordered and unordered list, link, link removal, glossification
* DO NOT INCLUDE: `<cite>` `<blockquote cite>` `<code>`, pull quotes, `<dl><dt><dd>`, or `<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-display-settings data-align data-caption data-embed-button>`

We're using `usa-measure` to control content wrapping.

#### DoR Tasks

- [x] Review items allowed in NCIDS Streamlined WYSIWYG
- [x] Content Model Drafted @andyvanavery31 
- [x] Content Model Reviewed
  - [x] Developer Review
  - [x] IA Review
- [x] Content Model Approved
  - [x]  Developer Approval
  - [x] IA Approval
- [x] Acceptance Criteria Drafted
  - [x] Analytics Acceptance Criteria Drafted
- [x] Scenarios created for YAML
- [ ] Acceptance Criteria Reviewed
  - [ ] Developer Review
  - [ ] IA Review
  - [ ] Analytics Review
- [ ] Acceptance Criteria Approved
  - [ ] Developer Approval
  - [ ] IA Approval
  - [ ] Analytics Approval

#### DoD Tasks

- [ ] Governance Drafted
- [ ] PR Review (see PR for review and comments)
- [ ] Client Review with Demo

### Technical Notes

**Content Model Links**
- [Paragraph Types update](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Paragraph-Types#ncids-mini-landing-page-components)
- [Block Types update](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Block-Types)
- New [NCIDS Content Block](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Block-Type-NCIDS-Content-Block) page

**Scenarios**
On the, "The Chronicles and Adventures...", page please add the following 2 content blocks. 
**Content Block 1**
```
<h2>Level 2 heading</h2>

<p><strong>More than 30 percent</strong> of all human cancers - <em>including 95 percent of pancreatic cancers and 45 percent of colorectal cancers</em> - are driven by <sup>mutations of the RAS</sup> family of genes. <sub>NCI established the RAS initiative in 2013 to explore</sub> innovative approaches for attacking the <s>proteins encoded by mutant forms of RAS genes</s> and to <strong><em>ultimately create effective, new therapies for RAS-related cancers.</em></strong></p>

<h3>This is an h3 test</h3>

<h4>Internal Link Testing</h4>
<!-- Make this a managed link in the WYSIWYG -->
<p>See <a href="/about-cancer">About Cancer</a> managed internal link.</p>
<p>See <a href="/about-cancer">About Cancer</a> internal link.</p>

<h5>Media Link Testing</h5>
<!-- Make this a managed link to the following cgov_file in the WYSIWYG -->
<p>See <a href="/research/progress/discovery/test-file-url">Test File Download</a> managed media link.</p>

<h6>External Link Testing</h6>
<p>See <a href="https://www.google.com/">Google</a> for external link.</p>

<h2>List Testing</h2>

<ul>
  <li>This is</li>
  <li>an unordered</li>
  <li>list of stuff</li>
</ul>
<ol>
  <li>This is</li>
  <li>>an ordered</li>
  <li>list of stuff</li>
</ul>
```

**Content Block 2**
```
This is a <a href="https://www.google.com">test link without a heading</a>.
```
--------------------------

### Analytics Requirements

For the following values in the acceptance criteria, we will utilize the default values, so no data attributes will need to be added within each link:

- location: 'Body'
- componentType: 'Content Block'
- pageRows: total number of rows on the page overall
- pageRowIndex: the index of the selected row from the link click
- pageRowCols: Default is 0
- pageRowColIndex: Default is 0
- containerItems: Default is 1
- containerItemIndex: Default is 1
- componentTheme: Default is 'Not Defined'
- componentVariant: Default is 'Not Defined'
- title: This value should match the nearest heading placed above the link within the content block. If there is no heading in the Content Block above the link, return 'Not Defined'.
- linkType: This should be set based on the type of link. The values will be:
   - `External` - This is for external links outside of the host web site. Conditions that match this:
      - If the url start with `http://` or `https://` AND the hostname does not match the current hostname in the browser. (e.g., the href is `https://www.cancer.gov/foo` and the visitor is on the web page `https://dceg.cancer.gov/bar`)
   - `Media` - This is for managed links in the WYSIWYG to media entities (e.g., files and infographics). Conditions that match this:
      - The `<a>` element has the `data-entity-type="media"` attribute.
   - `Internal` - This is for managed and unmanaged links within the same host. Conditions that match this:
      -  If the url starts with `http://` or `https://` AND the hostname match the current hostname in the browser. (e.g., the href is `https://www.cancer.gov/foo` and the visitor is on the web page `https://www.cancer.gov/bar`)
      - The URL does not specify a protocol (i.e. it does not start with `something:` and it does not match the criteria of `Media`.
   - `Email` - This is for mailto: links within the content. Conditions that match this:
      - If the URL starts with `mailto:`.
   - `Other` - This is for any URLs that have a protocol that we do not know, e.g., `tel:` or `javascript:`. Conditions that match this:
      - If none of the other above conditions were met.
- linkArea: This will pass 'Text' based on the link clicked.
- totalLinks: This will count the total number of links inside the Content Block.
- linkPosition: This will count the position of the link in the Content Block. 
- linkText: The text content of the link clicked

### Acceptance Criteria

**CMS User Experience**
```gherkin
Scenario: NCIDS Content Block paragraph can be added to Mini Landing Page content type
  Given the user is logged in as an editor 
    And the user is editing a "Mini Landing Page" content item
  When the user expands the dropdown to add a paragraph to Contents
  Then the dropdown displays "Add NCIDS Content Block"
  When the user selects "Add NCIDS Content Block"
  Then the fields for a "NCIDS Content Block" paragraph appear
    And the "Contents" field appears
    And the field is required
    And it is a streamlined WYSIWYG
    And the streamlined WYSIWYG will have headings (`h1`-`h6`), inline styles (bold, italics, strikethrough, superscript, subscript), remove format, ordered and unordered list, link, link removal, glossification
    And the streamlined WYSIWYG will NOT have `<cite>` `<blockquote cite>` `<code>`, pull quotes, `<dl><dt><dd>`, or `<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-display-settings data-align data-caption data-embed-button>`
  When a content editor adds anything that is not allowed in the streamlined WYSIWYG
  Then the added content that is not allowed will not be rendered
  When the content editor has added a link
    And the content editor uses the Remove Link function
  Then the link will be removed
  When the content editor has added formatting to text (e.g., bold, italics, strikethrough, superscript, or subscript)
    And the content editor uses the remove format function
  Then the formatting will be removed from the text 
```

**Front End User Experience**
```gherkin
Scenario: A site visitor goes to a Mini Landing Page that uses the NCIDS style and has an NCIDS Content Block on it
  Given the Mini Landing Page uses the NCIDS style
    And the Mini Landing Page has an NCIDS Content Block on it
    And the NCIDS Content Block has the text from the Scenarios section above
  When the visitor views the Mini Landing Page with the NCIDS style and NCIDS Content Block
  Then the visitor will see the text from the Scenarios section above
    And the formatting will be the same as the Scenarios section above
    And the text will appear within the width of the usa-measure
  When the user clicks on the internal link (About Cancer)
  Then they will be redirected to the About Cancer page
  When the user clicks on the external link (Google)
  Then they will be redirected to the Google homepage
```

**Analytics Requirements**
```gherkin
Scenario: Analytics events for the NCIDS Content Block on Mini Landing Pages using the NCIDS style
  Given a user is on a Mini Landing Page
    And the Mini Landing Page uses the NCIDS style
    And the Mini Landing Page has an NCIDS Content Block
  When the user clicks on a link in the NCIDS Content Block
  Then there will be an NCIDataLayer Other event
    And the event has the name "MLP:ContentBlock:LinkClick"
    And the event has the linkname "MLP:ContentBlock:LinkClick"
    And this event will have the following data
  | key | value |
    | location | body |
    | pageRows | This will return the number of rows on the page overall |
    | pageRowIndex | This will return the index of the row number on the page that was clicked |
    | pageRowCols | Default is the number of columns of components in the container |
    | pageRowColIndex | Default is the index of the component by columns in the container that was clicked |
    | containerItems | Default is number of components in the container that was clicked |
    | containerItemIndex | Default is the index of the component in the container that was clicked |
    | componentType | Default to 'Content Block' |
    | componentTheme | Default to 'Not Defined' |
    | linkText | Default to the text content of the link clicked. This will match the text of the link clicked |
    | componentVariant | Default is 'Not Defined' |
    | title | This value will reflect the nearest heading above the link within the content block. If there is no nearest heading above the link within the content block, set to 'Not Defined' |
    | linkType | This will be defined as 'Internal', 'External', 'Media', 'Email', or 'Other' based on the type of link in the code as specified in the Analytics Requirements section above |
    | linkArea | This will be set to 'Text' for all links in the Content Block |
    | totalLinks | This value will reflect the total number of links in the Content Block |
    | linkPosition | This will be a value equal to the position of the link in the Content Block counting sequentially in reading order from the beginning of the Content Block |
```

### Resources:
* [Figma](https://www.figma.com/design/GCOniv8OTyj9OC1udKwo6S/Mini-Landing-Pages?node-id=4132-35842&t=y2q8UjojxqvzHKv6-0)
* [R3 Design Scope](https://nih.sharepoint.com/:x:/r/sites/NCI-OCPLNewCgovUXDesignSystem/_layouts/15/Doc.aspx?sourcedoc=%7B8985BEA6-3D0D-4A03-8AB6-83D26B12AC10%7D&file=R3%20Design%20Scope.xlsx&action=default)

---

**Comment by welshja** [comment]
@bryanpizzillo - provided this list of elements currently allowed in the existing streamlined WYSIWYG editor:

<em> <strong> <cite> <blockquote cite> <code> <ul type> <ol type start> <li> <dl> <dt> <dd> <h2 id> <h3 id> <h4 id> <h5 id> <h6 id> <p> <a class href hreflang data-entity-substitution data-entity-type data-entity-uuid data-glossary-id title> <drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-display-settings data-align data-caption data-embed-button> <s> <sup> <sub> <h1>

**Comment by welshja** [comment]
Initial Notes from Chat with Adriana (will follow-up with dev team to review and update)

To be allowed: 
`<em><strong><ul type> <ol type start><li><h1><h2 id><h3 id><h4 id> <h5 id> <h6 id> <p> <s> <sup> <sub> <a` class href hreflang data-entity-substitution data-entity-type data-entity-uuid data-glossary-id title> 
<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-display-settings data-align data-caption data-embed-button> `

Need a bit more discussion on these:
`<dl><dt><dd>`
- Any example of where this is being used?

`<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-display-settings data-align data-caption data-embed-button>`
- Any example of where this is being used?

Would we be able to do overrides like pull quote and things like that? (I think our hope is no this would be a bit more locked down? But let's discuss.

To not be allowed:
`<cite> <blockquote cite><code>`

**Comment by bryanpizzillo** [comment]
@zhuomingao can answer any questions to where things might be used. 

However for `<drupal-entity>` that is for the Drupal embeds and you should not allow it. Also, it should never been allowed into a field using the streamlined editor - my guess it was a legacy migration thing.

Also, even if `dl/dt/dd` are used, they probably should not be either. These should only be used when doing a list of  definitions. Or things where you might have something like:
* **Thing** - description for thing.
* **Other Thing** - description for other thing.

**Comment by adrianacastaneda** [comment]
@bryanpizzillo I'll ask Min to run an impact report just incase for the dl/dt/dd tags. If we find something, we'll note those pages for when they're converted to the new template. 

But I agree. Let's move forward without not allowing dl/dt/dd tags

Re: <drupal-entity>, this doesn't have anything to do with adding managed links to a page, right? In other words, if we keep `data-entity-substitution data-entity-type data-entity-uuid` but remove `<drupal-entity>` we should still be able to add a link to page using it node/page title?

**Comment by bryanpizzillo** [comment]
> <a class href hreflang data-entity-substitution data-entity-type data-entity-uuid data-glossary-id title>

No, links are `<a class href hreflang data-entity-substitution data-entity-type data-entity-uuid data-glossary-id title>`. 

The embeds are probably because people had images embedded in intro text in Percussion, and after content was cleaned up no one entered a ticket to remove this. However, the buttons are not in the WYSIWYG, so you could only make it work by pasting the HTML in the source view. However, well, those types of things *HAVE* to happen. 🙄 So it would have been someone on our team doing that. I hope there are no `<drupal-entity>` instances where it is the streamlined WYSIWYG. 

Let me call this out specifically, but we should look across the current mini-landing content blocks that today use the full-HTML editor. So when looking for oddballs like <cite> <dt> <code> etc. You can assume that any `<drupal-entity>` should be turned into a 2 column...

**Comment by andyvanavery31** [comment]
Decision as of 18JUN2024: Pull quotes will not be utilized in the Streamlined WYSIWYG. Bringing source code into the Streamlined WYSIWYG would be against the goal of this component, so it will not be allowed.

**Comment by dev-rana-publicis** [comment]
### Approach

- [ ] Create Drupal Paragraph
    - Label: NCIDS Content Block
    - Machine name: `ncids_content_block`
    - Description: **Needs DOR**
    - [ ] Create Fields
        - **Need to confirm with product about these fields**
        - `field_content_heading`
        - `field_html_content`
    - [ ] Add to allowed types to MLP contents
- [ ] Create Drupal template for Content Block
    - `paragraph--ncids-content-block.html.twig`
- [ ] Create entrypoint for the paragraph
    - `cgdp-content-block`
- [ ] Create CSS for cgdp-content-block
- [ ] Analytics?
- [ ] Add content block to one of the Example YAML pages
    - **Needs DOR**

**Comment by bryanpizzillo** [comment]
@andyvanavery31 and @sarinapadilla I updated the analytics requirements based on a call with @welshja because I noticed the `linkType` was not caring about `Media` when taking over the code from @dev-rana-publicis. This lead to more questions and wording adjustment. (Content Blocks, as opposed to Raw HTML, do not allow content editors to set analytics parameters via data attributes.)

I will also be updating the scenario content to add some of the other use cases. (e.g. `Email`, `Media`, and `Other`.) Absolute URLs (e.g., `https://www.cancer.gov/foo`) can't be tested in yaml so this test case will have to be done manually and added to the CMS content creation tests if we want to automate the testing.
